### PR TITLE
Sundials/ML v5.8.0p0

### PR DESCRIPTION
**Pull request by tbrk**
Upgrade to the latest release of Sundials/ML which supports versions 2.7.0 to 5.8.0 inclusive of the Sundials suite of numerical solvers from LLNL.

### CODE REVIEW

**Comment by kit-ty-kate** [comment]
It looks like your makefile does not support parallel builds:
```
# File "arkode/arkode.mli", line 1:
# Error: Corrupted compiled interface
# lsolvers/sundials_NonlinearSolver_impl.cmi
# make[1]: *** [variables.mk:13: arkode/arkode.cmi] Error 2
# make[1]: Leaving directory '/home/opam/.opam/4.03/.opam-switch/build/sundialsml.5.8.0p0/src'
# make: *** [Makefile:8: all] Error 2
```
You can choose to disable parallel build by removing the `-j` argument or fix it first, force-push and update the checksum here.

**Comment by tbrk** [comment]
Thanks @kit-ty-kate. I believe the problem is due to an out of date `.depend`. I've patched for this problem and a number of others identified by opam-ci (using the docker images to test locally; they're really helpful). I've also updated the `conf-sundials` package since some of the build problems were caused by problems in the distro-specific package lists. Hopefully the tests will run better on this attempt.

**Comment by tbrk** [comment]
I've tried to mark `oraclelinux` as `x-ci-accept-failures`, but it does not seem to be working.
Otherwise, the parallel build failures are a mystery to me. It fails with
```
# File "ida/ida.mli", line 1:
# Error: Corrupted compiled interface
# lsolvers/sundials_NonlinearSolver_impl.cmi
```
But `src/.depend` contains:
```
ida/ida.cmi : \
    sundials/sundials_RealArray.cmi \
    lsolvers/sundials_NonlinearSolver.cmi \
    sundials/sundials_Logfile.cmi \
    lsolvers/sundials_LinearSolver.cmi \
    sundials/sundials.cmi \
    nvectors/nvector_serial.cmi \
    nvectors/nvector.cmi \
    ida/ida_impl.cmo

```
Have I missed something obvious?
I've been parallel building for months with `-j48` on debian and I have never had such problems.

**Comment by mseri** [comment]
I can send a separate PR to add upper bounds to zelus since sundials interface has changed

File: packages/sundialsml/sundialsml.5.8.0p0/opam [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+depends: [
+    "ocaml" {>= "4.03.0"}
+    "base-bigarray"
+    "ocamlfind" {build}
+    "conf-sundials" {>= "2" & build}
+]
+depopts: [
+    "mpi"
+]
+synopsis: "Interface to the Sundials suite of numerical solvers"
```
Comment (mseri) [inline packages/sundialsml/sundialsml.5.8.0p0/opam:44]:
```suggestion
synopsis: "Interface to the Sundials suite of numerical solvers"
conflicts: ["base-domains"]
```
It is not compatible with multicore:
```
# cc -I /home/opam/.opam/4.12/lib/ocaml  -O3 -Wall -Werror  -DNDEBUG=1 -fPIC -isystem /home/opam/.opam/4.12/lib/ocaml -isystem /usr/include/suitesparse/  -o nvectors/nvector_pthreads_ml.o -c nvectors/nvector_pthreads_ml.c
# In file included from sundials/sundials_ml.c:25:
# sundials/sundials_ml.c: In function 'sunml_sundials_malloc_value':
# sundials/sundials_ml.c:189:32: error: 'Caml_black' undeclared (first use in this function)
#   189 |     *block = Make_header(1, 0, Caml_black);
#       |                                ^~~~~~~~~~
# sundials/sundials_ml.c:189:32: note: each undeclared identifier is reported only once for each function it appears in
# sundials/sundials_ml.c: In function 'sunml_wrap_session_pointer':
# sundials/sundials_ml.c:225:35: error: assignment of member 'compare' in read-only object
#   225 |     Custom_ops_val(vmem)->compare = compare_session_pointers;
#       |                                   ^
# make[1]: *** [Makefile:118: sundials/sundials_ml.o] Error 1
# make[1]: *** Waiting for unfinished jobs....
# make[1]: Leaving directory '/home/opam/.opam/4.12/.opam-switch/build/sundialsml.5.8.0p0/src'
# make: *** [Makefile:8: all] Error 2
```

Comment (tbrk) [reply packages/sundialsml/sundialsml.5.8.0p0/opam:44]: Thanks for those suggestions @mseri. That's very helpful.
I'm not surprised that the library does not (yet) work against multicore OCaml. I'll have a look at this eventually.
I don't know what's going on with the problems on arm. That is surprising.

**Comment by tbrk** [comment]
I think I know what is causing the parallel build problems. I think it's because the `*_impl.ml` files do not have a corresponding `*_impl.mli` file. I'm going to add a rule tonight to generate these with `ocamlc -i` and then compile them to `.cmi` files (unless there is a better way of handling this situation).

File: packages/sundialsml/sundialsml.5.8.0p0/opam [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    "simulation"
+    "mathematics"
+    "science"
+]
+license: "BSD-3-Clause"
+build: [
+    ["./configure" "--stubdir=%{stublibs}%/"
+		   "--libdir=%{lib}%/"
+		   "--docdir=%{doc}%/sundialsml/"
+		   "--os-distribution=%{os-distribution}%"]
```
Comment (mseri) [inline packages/sundialsml/sundialsml.5.8.0p0/opam:23]: I tested this with macports and worked fine

Comment (tbrk) [reply packages/sundialsml/sundialsml.5.8.0p0/opam:23]: Great! Thanks for the confirmation @mseri.
I've just resubmitted an update for the parallel build problems. It also includes an attempted fix for the arm64 by compiling with `-fPIC` as suggested (at far, far right) by the error messages.
I'll try to look into the multicore build when I have to upgrade to Sundials 6.0.0, which will apparently be quite soon.

**Comment by tbrk** [comment]
The last iteration seems to have resolved nearly all the issues.
I don't know how to fix the oraclelinux build: installing `sundials-devel` seems to require the addition of the `epel-release` repository, which seems to involve manually editing a configuration file. I don't know why the exclusion commands in `opam` do not work.
I don't really understand what the `revdeps` mean or what is expected.

**Comment by tbrk** [comment]
I've pushed a commit to bound the versions on most of the `revdeps` (based on @mseri's comment). Please feel free to disregard it. The previous commit that just concerns `conf-sundials` and `sundialsml.5.8.0p0` seems to build well.

**Comment by mseri** [comment]
I don't think we need that upper bound. The failures are fine since we cannot control what system packages are installed

**Comment by tbrk** [comment]
I've reset to the original commit. OK for merge?

**Comment by mseri** [comment]
Yes, thanks a lot

**Comment by tbrk** [comment]
Thanks @mseri and @kit-ty-kate!

**Pull request merged**